samples/extensions/sparse_image/sparse_image.cpp

```diff
@@ -0,0 +1,1512 @@
+	pipeline_create_info.pStages                      = shader_stages.data();
+	pipeline_create_info.pVertexInputState            = &vertex_input;
+	pipeline_create_info.pInputAssemblyState          = &input_assembly;
+	pipeline_create_info.pRasterizationState          = &raster;
+	pipeline_create_info.pColorBlendState             = &blend;
+	pipeline_create_info.pMultisampleState            = &multisample;
+	pipeline_create_info.pViewportState               = &viewport;
+	pipeline_create_info.pDepthStencilState           = &depth_stencil;
+	pipeline_create_info.pDynamicState                = &dynamic;
+
+	VK_CHECK(vkCreateGraphicsPipelines(get_device().get_handle(), pipeline_cache, 1U, &pipeline_create_info, nullptr, &sample_pipeline));
+}
+
+void SparseImage::setup_camera()
+{
+	camera.type = vkb::CameraType::FirstPerson;
+	camera.set_perspective(fov_degrees, static_cast<float>(width) / static_cast<float>(height), 0.1f, 512.0f);
+	camera.set_rotation(glm::vec3(0.0f, 0.0f, 0.0f));
+	camera.set_translation(glm::vec3(0.0f, 0.0f, -50.0f));
+	camera.translation_speed = 3.0f;
+}
+
+/**
+ * 	@brief Copy a single-page, raw-pixel-data from the CPU memory into the stagging buffer.
+ */
+void SparseImage::copy_single_raw_data_block(uint8_t buffer[], const VkExtent2D blockDim, VkOffset2D offset, size_t stride)
+{
+	for (size_t row = 0U; row < blockDim.height; row++)
+	{
+		size_t position = (row + offset.y) * stride + offset.x * 4U;
+		memcpy(&buffer[row * blockDim.width * 4U], &virtual_texture.raw_data_image->get_data()[position], blockDim.width * 4U);
+	}
+}
+
+/**
+ * 	@brief Fill up the information on how the sparse image should be bound and call vkQueueBindSparse, based on the bind_list.
+ */
+void SparseImage::bind_sparse_image()
+{
+	virtual_texture.bind_list.sort();
+
+	for (size_t count = 0U; count < virtual_texture.num_pages; count++)
+	{
+		if (virtual_texture.bind_list.empty() == true)
+		{
+			virtual_texture.sparse_image_memory_bind[count].memory = VK_NULL_HANDLE;
+			continue;
+		}
+
+		size_t page_index = virtual_texture.bind_list.front();
+		if (count != page_index)
+		{
+			virtual_texture.sparse_image_memory_bind[count].memory = VK_NULL_HANDLE;
+			continue;
+		}
+
+		virtual_texture.bind_list.pop_front();
+		if (virtual_texture.page_table[page_index].bound == true)
+		{
+			continue;
+		}
+
+		if (virtual_texture.available_memory_index_list.empty() == true)
+		{
+			throw std::runtime_error("bind_sparse_image(): There is not enough memory available");
+		}
+
+		size_t memory_block_index = virtual_texture.available_memory_index_list.front();
+		virtual_texture.available_memory_index_list.pop_front();
+
+		size_t pageOffset = virtual_texture.page_size * memory_block_index;
+
+		virtual_texture.page_table[page_index].memory_index = memory_block_index;
+		virtual_texture.page_table[page_index].bound        = true;
+
+		virtual_texture.sparse_image_memory_bind[count].memory       = virtual_texture.texture_memory;
```

Comment:
Does the sparse image extension require the same memory for all binds? Or is it just to keep the memory management simple?

Reply:
No, the extension does not require the same memory for all binds. Actually, it is one of the advantages of using sparse binding to not be restricted by the continuous memory space requirements. Since currently the memory for the image is allocated statically, it felt natural to use a single block of memory.

Reply:
Yes, initially that was my idea to allocate on demand, however I faced some performance-related issues. I think it was quite a while ago and the the sample design is now way more reliable so I would try again with dynamic approach.